# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
data_path = "drive/MyDrive/Homeworks/supermarket_sales.csv"
df = pd.read_csv(data_path)

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [4]:
df.head(4)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [5]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [6]:
df[615:620]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Invoice ID               1000 non-null   object 
 1   Branch                   1000 non-null   object 
 2   City                     1000 non-null   object 
 3   Customer type            1000 non-null   object 
 4   Gender                   1000 non-null   object 
 5   Product line             1000 non-null   object 
 6   Unit price               1000 non-null   float64
 7   Quantity                 1000 non-null   int64  
 8   Tax 5%                   1000 non-null   float64
 9   Total                    1000 non-null   float64
 10  Date                     1000 non-null   object 
 11  Time                     1000 non-null   object 
 12  Payment                  1000 non-null   object 
 13  cogs                     1000 non-null   float64
 14  gross margin percentage  

In [8]:
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [9]:
percentiles = [0.01, 0.25, 0.50, 0.75, 0.99]
df['gross income'].describe(percentiles=percentiles)

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

75-й персентиль (gross income) = 22.45;
75% покупок мають gross income менше чи рівне 22.45;
25% покупок перевищують цю суму.


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [10]:
df ["City"].value_counts()

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64

In [11]:
df.groupby("City").size()

City
Mandalay     332
Naypyitaw    328
Yangon       340
dtype: int64

In [12]:
# checking unique cities
df.City.unique()

array(['Yangon', 'Naypyitaw', 'Mandalay'], dtype=object)

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [13]:
df ["City"].value_counts(normalize=True)*100

City
Yangon       34.0
Mandalay     33.2
Naypyitaw    32.8
Name: proportion, dtype: float64

In [14]:
df ["City"].value_counts(normalize=True)

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: proportion, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [15]:
yangon_mean = df[df['City']== 'Yangon']['gross income'].mean().round(2)
mandalay_mean = df[df['City']== 'Mandalay']['gross income'].mean().round(2)
print(yangon_mean)
print(mandalay_mean)

14.87
15.23


In [16]:
if yangon_mean > mandalay_mean:
  print("Yangon has a higher mean gross income")
elif yangon_mean < mandalay_mean:
  print ("Mandalay has a higher mean gross income")
else:
  print("Mean gross income is the same")

Mandalay has a higher mean gross income


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

In [17]:
df.columns

Index(['Invoice ID', 'Branch', 'City', 'Customer type', 'Gender',
       'Product line', 'Unit price', 'Quantity', 'Tax 5%', 'Total', 'Date',
       'Time', 'Payment', 'cogs', 'gross margin percentage', 'gross income',
       'Rating'],
      dtype='object')

### Hypothesis 1: Higher average unit price in the city
*   Rationale: if the product is sold at a higher price, while the costs remain similar, gross income will be higher
*   Testing: suggestion to compare average unit price across the cities, evaluate correlation between unit price and gross income
*   Additional data: income level of the population
*   Results: applying city specific pricing strategy; if the demand is not reduced then suggestion to test price increase in other cities

In [18]:
avg_price_per_city = df.groupby ('City')['Unit price'].mean().sort_values(ascending=False)
print('Average unit price per city:')
print(avg_price_per_city)

Average unit price per city:
City
Naypyitaw    56.609024
Mandalay     55.659277
Yangon       54.780853
Name: Unit price, dtype: float64


In [19]:
avg_gross_income_per_city = df.groupby('City')['gross income'].mean().sort_values(ascending=False)
print('Average gross income per city:')
print(avg_gross_income_per_city)

Average gross income per city:
City
Naypyitaw    16.052367
Mandalay     15.232024
Yangon       14.874001
Name: gross income, dtype: float64


In [20]:
correlation = df['Unit price'].corr(df['gross income'])
print('Correlation between unit price and gross income:', correlation)

Correlation between unit price and gross income: 0.633962088589069


There is a moderate positive correlation, higher unit price - higher income

### Hypothesis 2: Customers buy several items in one purchase

*   Rationale: if customers buy more units per transaction, the gross income also increases
*   Testing: compare average quantity per transaction across cities, analyze customer types as well as impact on purchase
*   Additional data: visitors count, demographic indicators
*   Results: where lower number of units identified per transaction, introduce promotions; encourage multi-item purchases





In [21]:
avg_quantity_per_city = df.groupby('City')['Quantity'].mean().sort_values(ascending=False)
print('Average quantity per city :')
print(avg_quantity_per_city)

Average quantity per city :
City
Naypyitaw    5.582317
Mandalay     5.481928
Yangon       5.467647
Name: Quantity, dtype: float64


In [22]:
avg_total_per_city = df.groupby('City')['Total'].mean().sort_values(ascending=False)
print('Average total per city:')
print(avg_total_per_city)

Average total per city:
City
Naypyitaw    337.099715
Mandalay     319.872506
Yangon       312.354031
Name: Total, dtype: float64


In [23]:
correlation = df['Quantity'].corr(df['gross income'])
print('Overall correlation between quantity and gross income:', correlation)

Overall correlation between quantity and gross income: 0.7055101859433067


There is a moderate positive correlation, higher quantity - higher income

In [24]:
unique_invoices = df['Invoice ID'].nunique()
print('Number of unique invoices is', unique_invoices)


Number of unique invoices is 1000


### Hypothesis 3: The city has a better (more profitable) product line
*   Rationale: if customers buy more expensive (high-margin products), the gross income will be higher
*   Testing: comparing the share of each product line across cities
*   Additional data: information on promotions, marketing campaigns, product lines data
*   Results: promote profitable product lines, optimize assortment accross stores, add high-margin products





In [25]:
total_gross_income = df.groupby(['City', 'Product line'])['gross income'].sum()
total_gross_income_sorted = total_gross_income.groupby(level=0, group_keys= False).apply(lambda x:x.sort_values(ascending=False))
print(total_gross_income_sorted)

City       Product line          
Mandalay   Sports and travel          951.8190
           Health and beauty          951.4600
           Home and lifestyle         835.6745
           Electronic accessories     811.9735
           Fashion accessories        781.5865
           Food and beverages         724.5185
Naypyitaw  Food and beverages        1131.7550
           Fashion accessories       1026.6700
           Electronic accessories     903.2845
           Health and beauty          791.2060
           Sports and travel          750.5680
           Home and lifestyle         661.6930
Yangon     Home and lifestyle        1067.4855
           Sports and travel          922.5095
           Electronic accessories     872.2435
           Food and beverages         817.2905
           Fashion accessories        777.7385
           Health and beauty          599.8930
Name: gross income, dtype: float64


In [26]:
gross_income_share = total_gross_income.groupby(level=0, group_keys=False).apply(lambda x: 100* x/x.sum()).round(2)
gross_income_share_sorted = gross_income_share.groupby(level=0,group_keys=False).apply(lambda x: x.sort_values())
print(gross_income_share_sorted)

City       Product line          
Mandalay   Food and beverages        14.33
           Fashion accessories       15.46
           Electronic accessories    16.06
           Home and lifestyle        16.52
           Health and beauty         18.81
           Sports and travel         18.82
Naypyitaw  Home and lifestyle        12.57
           Sports and travel         14.26
           Health and beauty         15.03
           Electronic accessories    17.16
           Fashion accessories       19.50
           Food and beverages        21.50
Yangon     Health and beauty         11.86
           Fashion accessories       15.38
           Food and beverages        16.16
           Electronic accessories    17.25
           Sports and travel         18.24
           Home and lifestyle        21.11
Name: gross income, dtype: float64


In [27]:
df.groupby('City')['gross income'].mean()

City
Mandalay     15.232024
Naypyitaw    16.052367
Yangon       14.874001
Name: gross income, dtype: float64

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [28]:
# Унікальні значення маржі
unique_margin_per_invoice = df.groupby('Invoice ID')['gross margin percentage'].nunique()
print(unique_margin_per_invoice)

Invoice ID
101-17-6199    1
101-81-4070    1
102-06-2002    1
102-77-2261    1
105-10-6182    1
              ..
894-41-5205    1
895-03-6665    1
895-66-0685    1
896-34-0956    1
898-04-2717    1
Name: gross margin percentage, Length: 1000, dtype: int64


In [29]:
if (unique_margin_per_invoice ==1).all():
  print ('All invoices have the same margin')
else:
  print('Gross margin varies')

All invoices have the same margin


Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [30]:
df_cities = df[df['City'].isin(['Yangon', 'Mandalay'])]
invoice_counts = df_cities.groupby(['City', 'Product line'])['Invoice ID'].nunique()
invoice_percent = invoice_counts.groupby(level=0, group_keys=False).apply(lambda x: x/x.sum())
print(invoice_percent)

City      Product line          
Mandalay  Electronic accessories    0.165663
          Fashion accessories       0.186747
          Food and beverages        0.150602
          Health and beauty         0.159639
          Home and lifestyle        0.150602
          Sports and travel         0.186747
Yangon    Electronic accessories    0.176471
          Fashion accessories       0.150000
          Food and beverages        0.170588
          Health and beauty         0.138235
          Home and lifestyle        0.191176
          Sports and travel         0.173529
Name: Invoice ID, dtype: float64


Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)